{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MHC allele protein sequence provider"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Protein fasta download URL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "test_protein_seq (__main__.MHCAlleleSeqProviderTest) ... "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein seq for BoLA-1*023:01:\n",
      "Loading from file datasets/mhcinfo/prot/BoLA/1.fa\n",
      "Protein seq for BoLA-2*012:01:\n",
      "Loading from file datasets/mhcinfo/prot/BoLA/2.fa\n",
      "Protein seq for BoLA-3*001:01:\n",
      "Loading from file datasets/mhcinfo/prot/BoLA/3.fa\n",
      "Protein seq for BoLA-3*002:01:\n",
      "Protein seq for BoLA-6*013:01:\n",
      "Loading from file datasets/mhcinfo/prot/BoLA/6.fa\n",
      "Protein seq for BoLA-6*041:01:\n",
      "Protein seq for BoLA-T2c:\n",
      "Loading from file datasets/mhcinfo/prot/BoLA/T2c.fa\n",
      "Protein seq for H2-Db:\n",
      "Loading from file datasets/mhcinfo/prot/H2/D.fa\n",
      "Protein seq for H2-Dd:\n",
      "Protein seq for H2-Kb:\n",
      "Loading from file datasets/mhcinfo/prot/H2/K.fa\n",
      "Protein seq for H2-Kd:\n",
      "Protein seq for H2-Kk:\n",
      "Protein seq for H2-Ld:\n",
      "Loading from file datasets/mhcinfo/prot/H2/L.fa\n",
      "Protein seq for HLA-A*01:01:\n",
      "Loading from file datasets/mhcinfo/prot/HLA/A.fa\n",
      "Protein seq for HLA-A*02:01:\n",
      "Protein seq for HLA-A*02:02:\n",
      "Protein seq for HLA-A*02:03:\n",
      "Protein seq for HLA-A*02:05:\n",
      "Protein seq for HLA-A*02:06:\n",
      "Protein seq for HLA-A*02:07:\n",
      "Protein seq for HLA-A*02:11:\n",
      "Protein seq for HLA-A*02:12:\n",
      "Protein seq for HLA-A*02:16:\n",
      "Protein seq for HLA-A*02:17:\n",
      "Protein seq for HLA-A*02:19:\n",
      "Protein seq for HLA-A*02:50:\n",
      "Protein seq for HLA-A*03:01:\n",
      "Protein seq for HLA-A*03:19:\n",
      "Protein seq for HLA-A*11:01:\n",
      "Protein seq for HLA-A*23:01:\n",
      "Protein seq for HLA-A*24:02:\n",
      "Protein seq for HLA-A*24:03:\n",
      "Protein seq for HLA-A*25:01:\n",
      "Protein seq for HLA-A*26:01:\n",
      "Protein seq for HLA-A*26:02:\n",
      "Protein seq for HLA-A*26:03:\n",
      "Protein seq for HLA-A*29:02:\n",
      "Protein seq for HLA-A*30:01:\n",
      "Protein seq for HLA-A*30:02:\n",
      "Protein seq for HLA-A*31:01:\n",
      "Protein seq for HLA-A*32:01:\n",
      "Protein seq for HLA-A*32:07:\n",
      "Protein seq for HLA-A*32:15:\n",
      "Protein seq for HLA-A*33:01:\n",
      "Protein seq for HLA-A*66:01:\n",
      "Protein seq for HLA-A*68:01:\n",
      "Protein seq for HLA-A*68:02:\n",
      "Protein seq for HLA-A*68:23:\n",
      "Protein seq for HLA-A*69:01:\n",
      "Protein seq for HLA-A*80:01:\n",
      "Protein seq for HLA-B*07:02:\n",
      "Loading from file datasets/mhcinfo/prot/HLA/B.fa\n",
      "Protein seq for HLA-B*08:01:\n",
      "Protein seq for HLA-B*08:02:\n",
      "Protein seq for HLA-B*08:03:\n",
      "Protein seq for HLA-B*14:01:\n",
      "Protein seq for HLA-B*14:02:\n",
      "Protein seq for HLA-B*15:01:\n",
      "Protein seq for HLA-B*15:02:\n",
      "Protein seq for HLA-B*15:03:\n",
      "Protein seq for HLA-B*15:09:\n",
      "Protein seq for HLA-B*15:17:\n",
      "Protein seq for HLA-B*18:01:\n",
      "Protein seq for HLA-B*27:05:\n",
      "Protein seq for HLA-B*27:20:\n",
      "Protein seq for HLA-B*35:01:\n",
      "Protein seq for HLA-B*35:03:\n",
      "Protein seq for HLA-B*37:01:\n",
      "Protein seq for HLA-B*38:01:\n",
      "Protein seq for HLA-B*39:01:\n",
      "Protein seq for HLA-B*40:01:\n",
      "Protein seq for HLA-B*40:02:\n",
      "Protein seq for HLA-B*40:13:\n",
      "Protein seq for HLA-B*42:01:\n",
      "Protein seq for HLA-B*44:02:\n",
      "Protein seq for HLA-B*44:03:\n",
      "Protein seq for HLA-B*45:01:\n",
      "Protein seq for HLA-B*46:01:\n",
      "Protein seq for HLA-B*48:01:\n",
      "Protein seq for HLA-B*51:01:\n",
      "Protein seq for HLA-B*53:01:\n",
      "Protein seq for HLA-B*54:01:\n",
      "Protein seq for HLA-B*57:01:\n",
      "Protein seq for HLA-B*57:03:\n",
      "Protein seq for HLA-B*58:01:\n",
      "Protein seq for HLA-B*58:02:\n",
      "Protein seq for HLA-B*73:01:\n",
      "Protein seq for HLA-B*81:01:\n",
      "Protein seq for HLA-B*83:01:\n",
      "Protein seq for HLA-C*03:03:\n",
      "Loading from file datasets/mhcinfo/prot/HLA/C.fa\n",
      "Protein seq for HLA-C*04:01:\n",
      "Protein seq for HLA-C*05:01:\n",
      "Protein seq for HLA-C*06:02:\n",
      "Protein seq for HLA-C*07:01:\n",
      "Protein seq for HLA-C*07:02:\n",
      "Protein seq for HLA-C*08:02:\n",
      "Protein seq for HLA-C*12:03:\n",
      "Protein seq for HLA-C*14:02:\n",
      "Protein seq for HLA-C*15:02:\n",
      "Protein seq for HLA-E*01:01:\n",
      "Loading from file datasets/mhcinfo/prot/HLA/E.fa\n",
      "Protein seq for HLA-E*01:03:\n",
      "Protein seq for Mamu-A1*001:01:\n",
      "Loading from file datasets/mhcinfo/prot/Mamu/A1.fa\n",
      "Protein seq for Mamu-A1*002:01:\n",
      "Protein seq for Mamu-A1*007:01:\n",
      "Protein seq for Mamu-A1*011:01:\n",
      "Protein seq for Mamu-A1*022:01:\n",
      "Protein seq for Mamu-A1*026:01:\n",
      "Protein seq for Mamu-A2*01:02:\n",
      "Loading from file datasets/mhcinfo/prot/Mamu/A2.fa\n",
      "Protein seq for Mamu-A7*01:03:\n",
      "Loading from file datasets/mhcinfo/prot/Mamu/A7.fa\n",
      "Protein seq for Mamu-B*001:01:\n",
      "Loading from file datasets/mhcinfo/prot/Mamu/B.fa\n",
      "Protein seq for Mamu-B*003:01:\n",
      "Protein seq for Mamu-B*008:01:\n",
      "Protein seq for Mamu-B*010:01:\n",
      "Protein seq for Mamu-B*017:01:\n",
      "Protein seq for Mamu-B*039:01:\n",
      "Protein seq for Mamu-B*052:01:\n",
      "Protein seq for Mamu-B*066:01:\n",
      "Protein seq for Mamu-B*084:01:\n",
      "Protein seq for Mamu-B*087:01:\n",
      "Protein seq for Patr-A*01:01:\n",
      "Loading from file datasets/mhcinfo/prot/Patr/A.fa\n",
      "Protein seq for Patr-A*03:01:\n",
      "Protein seq for Patr-A*04:01:\n",
      "Protein seq for Patr-A*07:01:\n",
      "Protein seq for Patr-A*09:01:\n",
      "Protein seq for Patr-B*01:01:\n",
      "Loading from file datasets/mhcinfo/prot/Patr/B.fa\n",
      "Protein seq for Patr-B*13:01:\n",
      "Protein seq for Patr-B*24:01:\n",
      "Protein seq for Rano-A1*b:\n",
      "Loading from file datasets/mhcinfo/prot/Rano/A1.fa\n",
      "Protein seq for SLA-1*04:01:\n",
      "Loading from file datasets/mhcinfo/prot/SLA/1.fa\n",
      "Protein seq for SLA-1*07:01:\n",
      "Protein seq for SLA-2*04:01:\n",
      "Loading from file datasets/mhcinfo/prot/SLA/2.fa\n",
      "Protein seq for SLA-3*04:01:\n",
      "Loading from file datasets/mhcinfo/prot/SLA/3.fa\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ok\n",
      "\n",
      "----------------------------------------------------------------------\n",
      "Ran 1 test in 0.253s\n",
      "\n",
      "OK\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<unittest.runner.TextTestResult run=1 errors=0 failures=0>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%run 'commons.ipynb'\n",
    "%run 'mhcinfo.ipynb'\n",
    "\n",
    "import os\n",
    "import urllib.request\n",
    "import ssl\n",
    "from IPython.core.debugger import Tracer\n",
    " \n",
    "class MHCAlleleSeqProvider(object):\n",
    "    class FastaSeqLoader(FastaSeqParseListener):\n",
    "        def __init__(self, seq_provider):\n",
    "            self._seq_provider = seq_provider\n",
    "            \n",
    "        def on_seq_read(self, header=None, seq=None):\n",
    "            allele_name = self._get_allele_name(header)  \n",
    "#             print('on_seq_read: header:%s, allele:%s, seq:%s' % (header, allele_name, seq))\n",
    "            self._seq_provider._set_allele_seq(allele_name, seq)\n",
    "            \n",
    "        def _get_allele_name(self, header):\n",
    "#             Tracer()()\n",
    "            allele_name = None\n",
    "            if header.startswith('HLA'):\n",
    "                tokens = header.split()\n",
    "                allele_name = 'HLA-' + tokens[1].strip()\n",
    "            else:\n",
    "                tokens = header.split(',')\n",
    "                allele_name = tokens[1].strip()\n",
    "            \n",
    "            allele_name = MHCAlleleName.std_name(allele_name)\n",
    "            return MHCAlleleName.parse(allele_name).format()\n",
    "                        \n",
    "    \n",
    "    SPECIES_FASTA_BASEURL_MAP ={\n",
    "        'Patr': 'https://www.ebi.ac.uk/ipd/mhc/group/NHP/download/Patr',\n",
    "        'Mamu': 'https://www.ebi.ac.uk/ipd/mhc/group/NHP/download/Mamu',\n",
    "        'Gogo': 'https://www.ebi.ac.uk/ipd/mhc/group/NHP/download/Mamu',\n",
    "        'Eqca': 'https://www.ebi.ac.uk/ipd/mhc/group/ELA/download/Eqca',\n",
    "        'SLA': 'https://www.ebi.ac.uk/ipd/mhc/group/SLA/download/Susc',\n",
    "        'BoLA': 'https://www.ebi.ac.uk/ipd/mhc/group/BoLA/download/BoLA',\n",
    "        'Rano': 'https://www.ebi.ac.uk/ipd/mhc/group/RT1/download/Rano'\n",
    "    }\n",
    "    \n",
    "    # Class variable\n",
    "    _allele_seq_map = {}\n",
    "            \n",
    "    def protein_seq(self, allele_name):\n",
    "        if allele_name not in self._allele_seq_map:\n",
    "            allele = MHCAlleleName.parse(allele_name)\n",
    "            self._load_protein_fasta_by_gene(species=allele.species, gene=allele.gene)\n",
    "        seq = self._allele_seq_map[allele_name]\n",
    "        return seq\n",
    "    \n",
    "            \n",
    "    def _load_protein_fasta_by_gene(self, species, gene):\n",
    "        fn = 'datasets/mhcinfo/prot/%s/%s.fa' % (species, gene)\n",
    "        url = self._get_fasta_url(species, gene)\n",
    "\n",
    "        if not os.path.exists(fn):\n",
    "            print('Downloading from %s' % url)\n",
    "            if url is None:\n",
    "                raise ValueError('Undefined URL for %s, %s' % (species, gene))\n",
    "            # Download and save as the file\n",
    "            RemoteUtils.download_to(url, decode='utf-8', fnout=fn)\n",
    "            \n",
    "        parser = FastaSeqParser()\n",
    "        parser.add_parse_listener(MHCAlleleSeqProvider.FastaSeqLoader(self))\n",
    "        \n",
    "        with open(fn, 'r') as fin:\n",
    "            print('Loading from file %s' % fn)\n",
    "            parser.parse(fin)\n",
    "    \n",
    "    def _set_allele_seq(self, allele, seq):\n",
    "        if allele not in self._allele_seq_map:\n",
    "            self._allele_seq_map[allele] = seq\n",
    "\n",
    "    def _get_fasta_url(self, species, gene):\n",
    "        url = None\n",
    "        if species in self.SPECIES_FASTA_BASEURL_MAP:\n",
    "            baseurl = self.SPECIES_FASTA_BASEURL_MAP[species]\n",
    "            url = '%s/%s?type=protein' % (baseurl, gene)\n",
    "        return url\n",
    "        \n",
    "###################################################\n",
    "from unittest import *\n",
    "\n",
    "class MHCAlleleSeqProviderTest(TestCase):\n",
    "    def setUp(self):\n",
    "        self.classI_alleles = [\n",
    "            'BoLA-1*023:01', 'BoLA-2*012:01', 'BoLA-3*001:01', 'BoLA-3*002:01',\n",
    "            'BoLA-6*013:01', 'BoLA-6*041:01', 'BoLA-T2c', 'H2-Db', 'H2-Dd',\n",
    "            'H2-Kb', 'H2-Kd', 'H2-Kk', 'H2-Ld', 'HLA-A*01:01', 'HLA-A*02:01',\n",
    "            'HLA-A*02:02', 'HLA-A*02:03', 'HLA-A*02:05', 'HLA-A*02:06',\n",
    "            'HLA-A*02:07', 'HLA-A*02:11', 'HLA-A*02:12', 'HLA-A*02:16',\n",
    "            'HLA-A*02:17', 'HLA-A*02:19', 'HLA-A*02:50', 'HLA-A*03:01',\n",
    "            'HLA-A*03:19', 'HLA-A*11:01', 'HLA-A*23:01', 'HLA-A*24:02',\n",
    "            'HLA-A*24:03', 'HLA-A*25:01', 'HLA-A*26:01', 'HLA-A*26:02',\n",
    "            'HLA-A*26:03', 'HLA-A*29:02', 'HLA-A*30:01', 'HLA-A*30:02',\n",
    "            'HLA-A*31:01', 'HLA-A*32:01', 'HLA-A*32:07', 'HLA-A*32:15',\n",
    "            'HLA-A*33:01', 'HLA-A*66:01', 'HLA-A*68:01', 'HLA-A*68:02',\n",
    "            'HLA-A*68:23', 'HLA-A*69:01', 'HLA-A*80:01', 'HLA-B*07:02',\n",
    "            'HLA-B*08:01', 'HLA-B*08:02', 'HLA-B*08:03', 'HLA-B*14:01',\n",
    "            'HLA-B*14:02', 'HLA-B*15:01', 'HLA-B*15:02', 'HLA-B*15:03',\n",
    "            'HLA-B*15:09', 'HLA-B*15:17', 'HLA-B*18:01', 'HLA-B*27:05',\n",
    "            'HLA-B*27:20', 'HLA-B*35:01', 'HLA-B*35:03', 'HLA-B*37:01',\n",
    "            'HLA-B*38:01', 'HLA-B*39:01', 'HLA-B*40:01', 'HLA-B*40:02',\n",
    "            'HLA-B*40:13', 'HLA-B*42:01', 'HLA-B*44:02', 'HLA-B*44:03',\n",
    "            'HLA-B*45:01', 'HLA-B*46:01', 'HLA-B*48:01', 'HLA-B*51:01',\n",
    "            'HLA-B*53:01', 'HLA-B*54:01', 'HLA-B*57:01', 'HLA-B*57:03',\n",
    "            'HLA-B*58:01', 'HLA-B*58:02', 'HLA-B*73:01', 'HLA-B*81:01',\n",
    "            'HLA-B*83:01', 'HLA-C*03:03', 'HLA-C*04:01', 'HLA-C*05:01',\n",
    "            'HLA-C*06:02', 'HLA-C*07:01', 'HLA-C*07:02', 'HLA-C*08:02',\n",
    "            'HLA-C*12:03', 'HLA-C*14:02', 'HLA-C*15:02', 'HLA-E*01:01',\n",
    "            'HLA-E*01:03', 'Mamu-A1*001:01', 'Mamu-A1*002:01', 'Mamu-A1*007:01',\n",
    "            'Mamu-A1*011:01', 'Mamu-A1*022:01', 'Mamu-A1*026:01',\n",
    "            'Mamu-A2*01:02', 'Mamu-A7*01:03', 'Mamu-B*001:01', 'Mamu-B*003:01',\n",
    "            'Mamu-B*008:01', 'Mamu-B*010:01', 'Mamu-B*017:01', 'Mamu-B*039:01',\n",
    "            'Mamu-B*052:01', 'Mamu-B*066:01', 'Mamu-B*084:01', 'Mamu-B*087:01',\n",
    "            'Patr-A*01:01', 'Patr-A*03:01', 'Patr-A*04:01', 'Patr-A*07:01',\n",
    "            'Patr-A*09:01', 'Patr-B*01:01', 'Patr-B*13:01', 'Patr-B*24:01',\n",
    "            'Rano-A1*b', 'SLA-1*04:01', 'SLA-1*07:01', 'SLA-2*04:01','SLA-3*04:01']\n",
    "        \n",
    "        self.classII_alleles = [\n",
    "            'HLA-DPA1*01:03/DPB1*02:01', 'HLA-DPA1*01:03/DPB1*04:01', 'HLA-DPA1*02:01/DPB1*01:01', \n",
    "            'HLA-DPA1*02:01/DPB1*05:01', 'HLA-DPA1*03:01/DPB1*04:02', 'HLA-DQA1*01:01/DQB1*05:01', \n",
    "            'HLA-DQA1*01:02/DQB1*06:02', 'HLA-DQA1*03:01/DQB1*03:02', 'HLA-DQA1*04:01/DQB1*04:02', \n",
    "            'HLA-DQA1*05:01/DQB1*02:01', 'HLA-DQA1*05:01/DQB1*03:01', 'HLA-DRB1*01:01', 'HLA-DRB1*03:01', \n",
    "            'HLA-DRB1*03:02', 'HLA-DRB1*04:01', 'HLA-DRB1*04:04', 'HLA-DRB1*04:05', 'HLA-DRB1*07:01', \n",
    "            'HLA-DRB1*08:02', 'HLA-DRB1*08:06', 'HLA-DRB1*08:13', 'HLA-DRB1*08:19', 'HLA-DRB1*09:01', \n",
    "            'HLA-DRB1*11:01', 'HLA-DRB1*12:01', 'HLA-DRB1*12:02', 'HLA-DRB1*13:02', 'HLA-DRB1*14:02', \n",
    "            'HLA-DRB1*14:04', 'HLA-DRB1*14:12', 'HLA-DRB1*15:01', 'HLA-DRB3*01:01', 'HLA-DRB3*03:01', \n",
    "            'HLA-DRB4*01:01', 'HLA-DRB5*01:01', 'H2-IAb', 'H2-IAd']        \n",
    "    \n",
    "    def test_protein_seq(self):\n",
    "        seq_provider = MHCAlleleSeqProvider()\n",
    "        for i, old_name in enumerate(self.classI_alleles):\n",
    "            std_name = MHCAlleleName.std_name(old_name)\n",
    "            print('Protein seq for %s:' % (std_name))\n",
    "            seq = seq_provider.protein_seq(std_name)\n",
    "            self.assertTrue(seq is not None)\n",
    "#             print('Seq: %s' % seq)\n",
    "            self.assertTrue(SeqUtils.is_all_amino_acids(seq))\n",
    "            \n",
    "############\n",
    "# suite = TestSuite()\n",
    "# suite.addTests(TestLoader().loadTestsFromTestCase(MHCAlleleSeqProviderTest))\n",
    "# TextTestRunner(verbosity=3).run(suite)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mouse MHC alleles\n",
    "- Two mouse MHC-II alleles: H2-IAb and H2-IAd\n",
    "\n",
    "<code>\\>sp|P14434|HA2B_MOUSE H-2 class II histocompatibility antigen, A-B alpha chain OS=Mus musculus GN=H2-Aa PE=1 SV=2\n",
    "MPRSRALILGVLALTTMLSLCGGEDDIEADHVGTYGISVYQSPGDIGQYTFEFDGDELFY\n",
    "VDLDKKETVWMLPEFGQLASFDPQGGLQNIAVVKHNLGVLTKRSNSTPATNEAPQATVFP\n",
    "KSPVLLGQPNTLICFVDNIFPPVINITWLRNSKSVADGVYETSFFVNRDYSFHKLSYLTF\n",
    "IPSDDDIYDCKVEHWGLEEPVLKHWEPEIPAPMSELTETVVCALGLSVGLVGIVVGTIFI\n",
    "IQGLRSGGTSRHPGPL\n",
    "\\>sp|P14483|HB2A_MOUSE H-2 class II histocompatibility antigen, A beta chain OS=Mus musculus GN=H2-Ab1 PE=1 SV=1\n",
    "MALQIPSLLLSAAVVVLMVLSSPGTEGGDSERHFVYQFMGECYFTNGTQRIRYVTRYIYN\n",
    "REEYVRYDSDVGEHRAVTELGRPDAEYWNSQPEILERTRAELDTVCRHNYEGPETHTSLR\n",
    "RLEQPNVVISLSRTEALNHHNTLVCSVTDFYPAKIKVRWFRNGQEETVGVSSTQLIRNGD\n",
    "WTFQVLVMLEMTPRRGEVYTCHVEHPSLKSPITVEWRAQSESAWSKMLSGIGGCVLGVIF\n",
    "LGLGLFIRHRSQKGPRGPPPAGLLQ\n",
    "\\>sp|P04228|HA2D_MOUSE H-2 class II histocompatibility antigen, A-D alpha chain OS=Mus musculus GN=H2-Aa PE=1 SV=1\n",
    "MPCSRALILGVLALNTMLSLCGGEDDIEADHVGFYGTTVYQSPGDIGQYTHEFDGDELFY\n",
    "VDLDKKKTVWRLPEFGQLILFEPQGGLQNIAAEKHNLGILTKRSNFTPATNEAPQATVFP\n",
    "KSPVLLGQPNTLICFVDNIFPPVINITWLRNSKSVTDGVYETSFLVNRDHSFHKLSYLTF\n",
    "IPSDDDIYDCKVEHWGLEEPVLKHWEPEIPAPMSELTETVVCALGLSVGLVGIVVGTIFI\n",
    "IQGLRSGGTSRHPGPL\n",
    "\\>sp|P01921|HB2D_MOUSE H-2 class II histocompatibility antigen, A-D beta chain OS=Mus musculus GN=H2-Ab1 PE=1 SV=1\n",
    "MALQIPSLLLSAAVVVLMVLSSPRTEGGNSERHFVVQFKGECYYTNGTQRIRLVTRYIYN\n",
    "REEYVRYDSDVGEYRAVTELGRPDAEYWNSQPEILERTRAEVDTACRHNYEGPETSTSLR\n",
    "RLEQPNVAISLSRTEALNHHNTLVCSVTDFYPAKIKVRWFRNGQEETVGVSSTQLIRNGD\n",
    "WTFQVLVMLEMTPHQGEVYTCHVEHPSLKSPITVEWRAQSESARSKMLSGIGGCVLGVIF\n",
    "LGLGLFIRHRSQKGPRGPPPAGLLQ</code>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# MOUSE_ALLELE_SEQ_MAP = {\n",
    "#     MHCAlleleName.H2_IAb: ('MPRSRALILGVLALTTMLSLCGGEDDIEADHVGTYGISVYQSPGDIGQYTFEFDGDELFYVDLDKKETVWMLPEFGQLASFDPQGGLQNIAVVKHNLGVLTKRSNSTPATNEAPQATVFPKSPVLLGQPNTLICFVDNIFPPVINITWLRNSKSVADGVYETSFFVNRDYSFHKLSYLTFIPSDDDIYDCKVEHWGLEEPVLKHWEPEIPAPMSELTETVVCALGLSVGLVGIVVGTIFIIQGLRSGGTSRHPGPL', 'MALQIPSLLLSAAVVVLMVLSSPGTEGGDSERHFVYQFMGECYFTNGTQRIRYVTRYIYNREEYVRYDSDVGEHRAVTELGRPDAEYWNSQPEILERTRAELDTVCRHNYEGPETHTSLRRLEQPNVVISLSRTEALNHHNTLVCSVTDFYPAKIKVRWFRNGQEETVGVSSTQLIRNGDWTFQVLVMLEMTPRRGEVYTCHVEHPSLKSPITVEWRAQSESAWSKMLSGIGGCVLGVIFLGLGLFIRHRSQKGPRGPPPAGLLQ'),\n",
    "#     MHCAlleleName.H2_IAd: ('MPCSRALILGVLALNTMLSLCGGEDDIEADHVGFYGTTVYQSPGDIGQYTHEFDGDELFYVDLDKKKTVWRLPEFGQLILFEPQGGLQNIAAEKHNLGILTKRSNFTPATNEAPQATVFPKSPVLLGQPNTLICFVDNIFPPVINITWLRNSKSVTDGVYETSFLVNRDHSFHKLSYLTFIPSDDDIYDCKVEHWGLEEPVLKHWEPEIPAPMSELTETVVCALGLSVGLVGIVVGTIFIIQGLRSGGTSRHPGPL', 'MALQIPSLLLSAAVVVLMVLSSPRTEGGNSERHFVVQFKGECYYTNGTQRIRLVTRYIYNREEYVRYDSDVGEYRAVTELGRPDAEYWNSQPEILERTRAEVDTACRHNYEGPETSTSLRRLEQPNVAISLSRTEALNHHNTLVCSVTDFYPAKIKVRWFRNGQEETVGVSSTQLIRNGDWTFQVLVMLEMTPHQGEVYTCHVEHPSLKSPITVEWRAQSESARSKMLSGIGGCVLGVIFLGLGLFIRHRSQKGPRGPPPAGLLQ')}\n",
    "\n",
    "# MOUSE_ALLELE_DOMAIN_OFFSET_MAP = {\n",
    "#     MHCAlleleName.H2_IAb: (26, 27),\n",
    "#     MHCAlleleName.H2_IAd: (26, 27)\n",
    "# }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MHC allele sequence provider "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# import warnings\n",
    "\n",
    "\n",
    "# class MHCAlleleSeqProvider(object):\n",
    "    \n",
    "#     def allele_seq(self, allele, domain_offset=True):\n",
    "#         if allele.startswith(MHCAlleleName.HLA):\n",
    "#             if MHCAlleleName.is_hla_classI(allele):\n",
    "#                 return self._get_hla_psa(allele).allele_seq(allele, domain_offset=domain_offset)\n",
    "#             else: # HLA class II\n",
    "#                 alleleA, alleleB = MHCAlleleName.split_hla2(allele)\n",
    "#                 seqA = self._get_hla_psa(alleleA).allele_seq(alleleA, domain_offset=domain_offset)\n",
    "#                 seqB = self._get_hla_psa(alleleB).allele_seq(alleleB, domain_offset=domain_offset)\n",
    "#                 return seqA, seqB\n",
    "                \n",
    "#         elif allele.startswith(MHCAlleleName.H2): # mouse allele\n",
    "#             seqA, seqB = MOUSE_ALLELE_SEQ_MAP[allele]\n",
    "#             offsetA = offsetB = 0\n",
    "#             if domain_offset:\n",
    "#                 offsetA, offsetB = MOUSE_ALLELE_DOMAIN_OFFSET_MAP[allele]\n",
    "#             return seqA[offsetA:], seqB[offsetB:]\n",
    "#         else:\n",
    "#             raise ValueError('Unknown allele: %s' % allele)\n",
    "                    \n",
    "#     def _get_hla_psa(self, allele):\n",
    "#         hlagene = MHCAlleleName.sub_name(allele, level=1)\n",
    "#         return IMGTHLAProteinSequenceAlignment.get_psa(hlagene)\n",
    "        \n",
    "# #         hlagene = MHCAlleleName.sub_name(allele, level=1)\n",
    "# #         psa = MHCAlleleSeqProvider.get_psa(hlagene)\n",
    "# #         seq = psa.allele_seq(allele)\n",
    "# #         offset = MHCAlleleSeqProvider.DOMAIN_OFFSET_MAP[hlagene] if use_offset else 0\n",
    "# #         return seq[offset:]\n",
    "            \n",
    "\n",
    "# #################################\n",
    "# from unittest import *\n",
    "\n",
    "# class MHCAlleleSeqProviderTest(TestCase):\n",
    "#     def setUp(self):\n",
    "#         self.seq_provider = MHCAlleleSeqProvider()\n",
    "#         self.hlaI_alleles =  [\n",
    "#             'HLA-A*01:01', 'HLA-A*02:01', 'HLA-A*02:02', 'HLA-A*02:03', 'HLA-A*02:05', 'HLA-A*02:06', 'HLA-A*02:07', \n",
    "#             'HLA-A*02:11', 'HLA-A*02:12', 'HLA-A*02:16', 'HLA-A*02:17', 'HLA-A*02:19', 'HLA-A*02:50', 'HLA-A*03:01', \n",
    "#             'HLA-A*03:19', 'HLA-A*11:01', 'HLA-A*23:01', 'HLA-A*24:02', 'HLA-A*24:03', 'HLA-A*25:01', 'HLA-A*26:01', \n",
    "#             'HLA-A*26:02', 'HLA-A*26:03', 'HLA-A*29:02', 'HLA-A*30:01', 'HLA-A*30:02', 'HLA-A*31:01', 'HLA-A*32:01', \n",
    "#             'HLA-A*32:07', 'HLA-A*32:15', 'HLA-A*33:01', 'HLA-A*66:01', 'HLA-A*68:01', 'HLA-A*68:02', 'HLA-A*68:23', \n",
    "#             'HLA-A*69:01', 'HLA-A*80:01', 'HLA-B*07:02', 'HLA-B*08:01', 'HLA-B*08:02', 'HLA-B*08:03', 'HLA-B*14:01', \n",
    "#             'HLA-B*14:02', 'HLA-B*15:01', 'HLA-B*15:02', 'HLA-B*15:03', 'HLA-B*15:09', 'HLA-B*15:17', 'HLA-B*18:01', \n",
    "#             'HLA-B*27:05', 'HLA-B*27:20', 'HLA-B*35:01', 'HLA-B*35:03', 'HLA-B*37:01', 'HLA-B*38:01', 'HLA-B*39:01', \n",
    "#             'HLA-B*40:01', 'HLA-B*40:02', 'HLA-B*40:13', 'HLA-B*42:01', 'HLA-B*44:02', 'HLA-B*44:03', 'HLA-B*45:01', \n",
    "#             'HLA-B*45:06', 'HLA-B*46:01', 'HLA-B*48:01', 'HLA-B*51:01', 'HLA-B*53:01', 'HLA-B*54:01', 'HLA-B*57:01', \n",
    "#             'HLA-B*57:03', 'HLA-B*58:01', 'HLA-B*58:02', 'HLA-B*73:01', 'HLA-B*81:01', 'HLA-B*83:01', 'HLA-C*01:02', \n",
    "#             'HLA-C*01:03', 'HLA-C*02:02']\n",
    "#         self.hlaII_alleles = ['HLA-DPA1*01:03/DPB1*02:01', 'HLA-DPA1*01:03/DPB1*04:01', 'HLA-DPA1*02:01/DPB1*01:01', \n",
    "#                               'HLA-DPA1*02:01/DPB1*05:01', 'HLA-DPA1*03:01/DPB1*04:02', 'HLA-DQA1*01:01/DQB1*05:01', \n",
    "#                               'HLA-DQA1*01:02/DQB1*06:02', 'HLA-DQA1*03:01/DQB1*03:02', 'HLA-DQA1*04:01/DQB1*04:02', \n",
    "#                               'HLA-DQA1*05:01/DQB1*02:01', 'HLA-DQA1*05:01/DQB1*03:01', 'HLA-DRB1*01:01', \n",
    "#                               'HLA-DRB1*03:01', 'HLA-DRB1*03:02', 'HLA-DRB1*04:01', 'HLA-DRB1*04:04', 'HLA-DRB1*04:05', \n",
    "#                               'HLA-DRB1*07:01', 'HLA-DRB1*08:02', 'HLA-DRB1*08:06', 'HLA-DRB1*08:13', 'HLA-DRB1*08:19', \n",
    "#                               'HLA-DRB1*09:01', 'HLA-DRB1*11:01', 'HLA-DRB1*12:01', 'HLA-DRB1*12:02', 'HLA-DRB1*13:02', \n",
    "#                               'HLA-DRB1*14:02', 'HLA-DRB1*14:04', 'HLA-DRB1*14:12', 'HLA-DRB1*15:01', 'HLA-DRB3*01:01', \n",
    "#                               'HLA-DRB3*03:01', 'HLA-DRB4*01:01', 'HLA-DRB5*01:01']\n",
    "#         self.mouse_alleles = ['H2-IAb', 'H2-IAd']\n",
    "    \n",
    "#     def test_allele_seq_for_hlaI(self):\n",
    "#         for allele in self.hlaI_alleles:\n",
    "#             seq = self.seq_provider.allele_seq(allele)\n",
    "#             self.assertTrue(seq is not None)\n",
    "#             self.assertTrue(len(seq) > 0)\n",
    "#             self.assertTrue(SeqUtils.is_all_amino_acids(seq, allow_dummy=True))\n",
    "\n",
    "#     def test_allele_seq_for_hlaII(self):\n",
    "#         for allele in self.hlaII_alleles:\n",
    "#             alpha_seq, beta_seq = self.seq_provider.allele_seq(allele)\n",
    "#             self.assertTrue(alpha_seq is not None)\n",
    "#             self.assertTrue(len(alpha_seq) > 0)\n",
    "#             self.assertTrue(SeqUtils.is_all_amino_acids(alpha_seq, allow_dummy=True))\n",
    "            \n",
    "#             self.assertTrue(beta_seq is not None)\n",
    "#             self.assertTrue(len(beta_seq) > 0)\n",
    "#             self.assertTrue(SeqUtils.is_all_amino_acids(beta_seq, allow_dummy=True))\n",
    "    \n",
    "#     def test_allele_seq_for_mouse_alleles(self):\n",
    "#         for allele in self.mouse_alleles:\n",
    "#             alpha_seq, beta_seq = self.seq_provider.allele_seq(allele)\n",
    "#             self.assertTrue(alpha_seq is not None)\n",
    "#             self.assertTrue(len(alpha_seq) > 0)\n",
    "#             self.assertTrue(SeqUtils.is_all_amino_acids(alpha_seq, allow_dummy=True))\n",
    "            \n",
    "#             self.assertTrue(beta_seq is not None)\n",
    "#             self.assertTrue(len(beta_seq) > 0)\n",
    "#             self.assertTrue(SeqUtils.is_all_amino_acids(beta_seq, allow_dummy=True))\n",
    "        \n",
    "# # ###########\n",
    "# # suite = TestSuite()\n",
    "# # suite.addTests(TestLoader().loadTestsFromTestCase(MHCAlleleSeqProviderTest))\n",
    "# # TextTestRunner(verbosity=3).run(suite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IKEEHVIIQAEFYLNPDQSGEFMFDFDGDEIFHVDMAKKETVWRLEEFGRFASFEAQGALANIAVDKANLEIMTKRSNYTPITNVPPEVTVLTNSPVELREPNVLICFIDKFTPPVVNVTWLRNGKPVTTGVSETVFLPREDHLFRKFHYLPFLPSTEDVYDCRVEHWGLDEPLLKHWEFDAPSPLPETTENVVCALGLTVGLVGIIIGTIFIIKGVRKSNAAERRGPL\n",
      "QFFFRFASFGANVDNI\n",
      "QFFFRFASFGANVDNI 16\n",
      "GDTRPRFLWQLKFECHFFNGTERVRLLERCIYNQEESVRFDSDVGEYRAVTELGRPDAEYWNSQKDLLEQRRAAVDTYCRHNYGVGESFTVQRRVEPKVTVYPSKTQPLQHHNLLVCSVSGFYPGSIEVRWFRNGQEEKAGVVSTGLIQNGDWTFQTLVMLETVPRSGEVYTCQVEHPSVTSPLTVEWRARSESAQSKMLSGVGGFVLGLLFLGAGLFIYFRNQKGHSGLQPTGFLS\n",
      "LFPDYWQRYHNV\n",
      "LFPDYWQRYHNV 12\n"
     ]
    }
   ],
   "source": [
    "# seq_provider = MHCAlleleSeqProvider()\n",
    "# allele = 'HLA-DRB1*01:01'\n",
    "# seqA, seqB = seq_provider.allele_seq(allele)\n",
    "# seqA = np.array(list(seqA))\n",
    "# seqB = np.array(list(seqB))\n",
    "\n",
    "# posA = [9, 22, 24, 32, 50, 51, 52, 53, 54, 58, 59, 62, 65, 66, 69, 72]\n",
    "# refA = 'QFFFRFASFGANVDNI'\n",
    "\n",
    "# posB =  [11, 13, 56, 57, 60, 61, 70, 71, 78, 81, 82, 85]\n",
    "# refB = 'LFPDYWQRYHNV'\n",
    "\n",
    "# posA = np.subtract(posA, 1)\n",
    "# posB = np.subtract(posB, 1)\n",
    "\n",
    "# print(''.join(seqA))\n",
    "# print(''.join(seqA[posA]))\n",
    "# print(refA, len(refA))\n",
    "# print(''.join(seqB))\n",
    "# print(''.join(seqB[posB]))\n",
    "# print(refB, len(refB))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "H-2-IAb\tYSYFLASGGQVVHVLYFGYTYHDIRTETVHGPHT\n",
    "H-2-IAd\tYTYHLILGGQAEHILVFGLTYYDIRTETAHGPST\n",
    "HLA-DPA10103-DPB10201\tYAFFMFSGGAILNTLFGQFEYFDIEEVRMHLGMT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nimport numpy as np\\n\\nallele = 'HLA-DPA1*01:03/DPB1*01:01'\\nmhcseq_provider = MHCAlleleSeqProvider()\\nalpha_seq, beta_seq = mhcseq_provider.allele_seq(allele)\\nbeta_seq = beta_seq[:22] + Const.DUMMY_AA + Const.DUMMY_AA + beta_seq[22:]\\n\\nalpha_seq = np.array(list(alpha_seq))\\nbeta_seq = np.array(list(beta_seq))\\n\\nrefseq = 'YAFFMFSGGAILNTLYGQFEYFAIEKVRVHLDVT'\\nposA = np.subtract([9, 11, 22, 24, 31, 52, 53, 58, 59, 61, 65, 66, 68, 72, 73], 1)\\nposB = np.subtract([9, 11, 13, 26, 28, 30, 47, 57, 67, 70, 71, 74, 77, 78, 81, 85, 86, 89 ,90], 1)\\n\\nprint('Allele:', allele)\\nprint('Alpha:', ''.join(alpha_seq))\\nprint('ref:', refseq[:15])\\nprint('chk:', ''.join(alpha_seq[posA]))\\n\\nprint('Beta:', ''.join(beta_seq))\\nprint('ref:', refseq[15:])\\nprint('chk:', ''.join(beta_seq[posB]))\\n\\n# H-2-IAb\\tYSYFLASGGQVVHVLYFGYTYHDIRTETVHGPHT\\n# H-2-IAd\\tYTYHLILGGQAEHILVFGLTYYDIRTETAHGPST\\n\\n# allele = 'H2-IAb'\\n# alpha_seq, beta_seq = mhcseq_provider.allele_seq(allele)\\n# alpha_seq = np.array(list(alpha_seq))\\n# beta_seq = np.array(list(beta_seq))\\n\\n# refseq = 'YSYFLASGGQVVHVLYFGYTYHDIRTETVHGPHT'\\n# posA = np.subtract([9, 11, 22, 24, 31, 52, 53, 58, 59, 61, 65, 66, 68, 72, 73], 1)\\n# posB = np.subtract([9, 11, 13, 26, 28, 30, 47, 57, 67, 70, 71, 74, 77, 78, 81, 85, 86, 89 ,90], 1)\\n\\n# print('Allele:', allele)\\n# print('Alpha:', ''.join(alpha_seq))\\n# print('ref:', refseq[:15])\\n# print('chk:', ''.join(alpha_seq[posA]))\\n\\n# print('Beta:', ''.join(beta_seq))\\n# print('ref:', refseq[15:])\\n# print('chk:', ''.join(beta_seq[posB]))\\n\\n# allele = 'H2-IAd'\\n# alpha_seq, beta_seq = mhcseq_provider.allele_seq(allele)\\n# alpha_seq = np.array(list(alpha_seq))\\n# beta_seq = np.array(list(beta_seq))\\n\\n# refseq = 'YTYHLILGGQAEHILVFGLTYYDIRTETAHGPST'\\n# posA = np.subtract([9, 11, 22, 24, 31, 52, 53, 58, 59, 61, 65, 66, 68, 72, 73], 1)\\n# posB = np.subtract([9, 11, 13, 26, 28, 30, 47, 57, 67, 70, 71, 74, 77, 78, 81, 85, 86, 89 ,90], 1)\\n\\n# print('Allele:', allele)\\n# print('Alpha:', ''.join(alpha_seq))\\n# print('ref:', refseq[:15])\\n# print('chk:', ''.join(alpha_seq[posA]))\\n\\n# print('Beta:', ''.join(beta_seq))\\n# print('ref:', refseq[15:])\\n# print('chk:', ''.join(beta_seq[posB]))\\n\\n\""
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import numpy as np\n",
    "\n",
    "# allele = 'HLA-DPA1*01:03/DPB1*01:01'\n",
    "# mhcseq_provider = MHCAlleleSeqProvider()\n",
    "# alpha_seq, beta_seq = mhcseq_provider.allele_seq(allele)\n",
    "# beta_seq = beta_seq[:22] + Const.DUMMY_AA + Const.DUMMY_AA + beta_seq[22:]\n",
    "\n",
    "# alpha_seq = np.array(list(alpha_seq))\n",
    "# beta_seq = np.array(list(beta_seq))\n",
    "\n",
    "# refseq = 'YAFFMFSGGAILNTLYGQFEYFAIEKVRVHLDVT'\n",
    "# posA = np.subtract([9, 11, 22, 24, 31, 52, 53, 58, 59, 61, 65, 66, 68, 72, 73], 1)\n",
    "# posB = np.subtract([9, 11, 13, 26, 28, 30, 47, 57, 67, 70, 71, 74, 77, 78, 81, 85, 86, 89 ,90], 1)\n",
    "\n",
    "# print('Allele:', allele)\n",
    "# print('Alpha:', ''.join(alpha_seq))\n",
    "# print('ref:', refseq[:15])\n",
    "# print('chk:', ''.join(alpha_seq[posA]))\n",
    "\n",
    "# print('Beta:', ''.join(beta_seq))\n",
    "# print('ref:', refseq[15:])\n",
    "# print('chk:', ''.join(beta_seq[posB]))\n",
    "\n",
    "# H-2-IAb\tYSYFLASGGQVVHVLYFGYTYHDIRTETVHGPHT\n",
    "# H-2-IAd\tYTYHLILGGQAEHILVFGLTYYDIRTETAHGPST\n",
    "\n",
    "# allele = 'H2-IAb'\n",
    "# alpha_seq, beta_seq = mhcseq_provider.allele_seq(allele)\n",
    "# alpha_seq = np.array(list(alpha_seq))\n",
    "# beta_seq = np.array(list(beta_seq))\n",
    "\n",
    "# refseq = 'YSYFLASGGQVVHVLYFGYTYHDIRTETVHGPHT'\n",
    "# posA = np.subtract([9, 11, 22, 24, 31, 52, 53, 58, 59, 61, 65, 66, 68, 72, 73], 1)\n",
    "# posB = np.subtract([9, 11, 13, 26, 28, 30, 47, 57, 67, 70, 71, 74, 77, 78, 81, 85, 86, 89 ,90], 1)\n",
    "\n",
    "# print('Allele:', allele)\n",
    "# print('Alpha:', ''.join(alpha_seq))\n",
    "# print('ref:', refseq[:15])\n",
    "# print('chk:', ''.join(alpha_seq[posA]))\n",
    "\n",
    "# print('Beta:', ''.join(beta_seq))\n",
    "# print('ref:', refseq[15:])\n",
    "# print('chk:', ''.join(beta_seq[posB]))\n",
    "\n",
    "# allele = 'H2-IAd'\n",
    "# alpha_seq, beta_seq = mhcseq_provider.allele_seq(allele)\n",
    "# alpha_seq = np.array(list(alpha_seq))\n",
    "# beta_seq = np.array(list(beta_seq))\n",
    "\n",
    "# refseq = 'YTYHLILGGQAEHILVFGLTYYDIRTETAHGPST'\n",
    "# posA = np.subtract([9, 11, 22, 24, 31, 52, 53, 58, 59, 61, 65, 66, 68, 72, 73], 1)\n",
    "# posB = np.subtract([9, 11, 13, 26, 28, 30, 47, 57, 67, 70, 71, 74, 77, 78, 81, 85, 86, 89 ,90], 1)\n",
    "\n",
    "# print('Allele:', allele)\n",
    "# print('Alpha:', ''.join(alpha_seq))\n",
    "# print('ref:', refseq[:15])\n",
    "# print('chk:', ''.join(alpha_seq[posA]))\n",
    "\n",
    "# print('Beta:', ''.join(beta_seq))\n",
    "# print('ref:', refseq[15:])\n",
    "# print('chk:', ''.join(beta_seq[posB]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IEADHVGFYGTTVYQSPGDIGQYTHEFDGDELFYVDLDKKKTVWRLPEFGQLILFEPQGGLQNIAAEKHNLGILTKRSNFTPATNEAPQATVFPKSPVLLGQPNTLICFVDNIFPPVINITWLRNSKSVTDGVYETSFLVNRDHSFHKLSYLTFIPSDDDIYDCKVEHWGLEEPVLKHWEPEIPAPMSELTETVVCALGLSVGLVGIVVGTIFIIQGLRSGGTSRHPGPL\n",
      "IEADHVGFYTTVYQSPGDIGQYTHEFDGDELFYVDLDKKKTVWRLPEFGQLILFEPQGGLQNIAAEKHNLGILTKRSNFTPATNEAPQATVFPKSPVLLGQPNTLICFVDNIFPPVINITWLRNSKSVTDGVYETSFLVNRDHSFHKLSYLTFIPSDDDIYDCKVEHWGLEEPVLKHWEPEIPAPMSELTETVVCALGLSVGLVGIVVGTIFIIQGLRSGGTSRHPGPL\n",
      "QRFLERYIYNREEFARFDSDVGEFRAVTELGRPAAEYWNSQKDILEEKRAVPDRMCRHNYELGGPMTLQRRVQPRVNVSPSKKGPLQHHNLLVCHVTDFYPGSIQVRWFLNGQEETAGVVSTNLIRNGDWTFQILVMLEMTPQQGDVYTCQVEHTSLDSPVTVEWKAQSDSARSKTLTGAGGFVLGLIICGVGIFMHRRSKKVQRGSA\n",
      "RATPENYLFQGRQECYAFNGTQXXRFLERYIYNREEFARFDSDVGEFRAVTELGRPAAEYWNSQKDILEEKRAVPDRMCRHNYELGGPMTLQRRVQPRVNVSPSKKGPLQHHNLLVCHVTDFYPGSIQVRWFLNGQEETAGVVSTNLIRNGDWTFQILVMLEMTPQQGDVYTCQVEHTSLDSPVTVEWKAQSDSARSKTLTGAGGFVLGLIICGVGIFMHRRSKKVQRGSA\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "allele = 'H2-IAd'\n",
    "alpha_seq, beta_seq = mhcseq_provider.allele_seq(allele)\n",
    "print(alpha_seq)\n",
    "newseq = alpha_seq[:9] + alpha_seq[10:]\n",
    "print(newseq)\n",
    "\n",
    "allele = 'HLA-DPA1*01:03/DPB1*04:01'\n",
    "alpha_seq, beta_seq = mhcseq_provider.allele_seq(allele)\n",
    "print(beta_seq[21:])\n",
    "newseq = beta_seq[:22] + Const.DUMMY_AA + Const.DUMMY_AA + beta_seq[22:]\n",
    "print(newseq)\n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  },
  "latex_envs": {
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 0
  },
  "nav_menu": {},
  "toc": {
   "base_numbering": 1,
   "nav_menu": {
    "height": "156px",
    "width": "252px"
   },
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "1141px",
    "left": "0px",
    "right": "966px",
    "top": "107px",
    "width": "225px"
   },
   "toc_section_display": "block",
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
